### ISSUE TITLE
Metadynamics on custom CV

### ISSUE THREAD

**Issue by bodhivani** [openmm/openmm#4017]
tl;dr - need to run metadynamics on 2 CVs that are linear combinations of atom-distances.
----
We've been trying to make openMM our primary MD engine, but I've been running into issues with running metadynamics.

I have 2 order parameters that are linear combinations of atom-distances, so I need to run metadynamics on a custom collective variable that takes in multiple x,y,z coordinates. I can't really find documentation that does this. I've written a basic code based on this: https://github.com/openmm/openmm/issues/2126, but I don't know how to define my custom CV.

If this cannot be done, I am going to try to use a plumed patch on OpenMM, although I haven't yet been able to successfully install the patch correctly, and I am also worried about the overhead of using plumed slowing everything down too much.

Is there some documentation page/resource than can help with this? Alternately, has someone done this already, and could guide me?

Thanks!

**Comment by peastman** [comment]
That should be possible.  Can you give details on exactly how you define the CVs?  How many distances are included in each CV?  Is the set of distances fixed or can it change with time?

**Comment by bodhivani** [comment]
Thank you for your quick response! The set of distances is fixed, I'm aiming for including ~14 but might decrease depending on stability of the simulation.

The CV is the latent space from a past-future information bottleneck with a linear encoder, so my CVs are just going to be

given a list of pairs of atoms,
$\sum_i c_i d_i$
where d_i = distance between the ith pair of atoms.

**Comment by peastman** [comment]
You can implement that with a CustomBondForce:

```python
cv = CustomBondForce("c*r")
cv.addPerBondParameter("c")
```

Then for each distance you want to include, call

```python
cv.addBond(i, j, [c])
```

where `i` and `j` are the indices of the two atoms and `c` is the weight to multiply the distance by.  The result will be a weighted sum of the distances.

You'll create two different CustomBondForces, one for each CV, containing different lists of atom pairs and weights.

**Comment by bodhivani** [comment]
Just implemented, so far works! Thank you so much-- it's a very usable and comprehensive code!

**Comment by peastman** [comment]
Great!  Is it ok to close this issue?

**Comment by peastman** [comment]
If you want to calculate an integral over the bias function, you'll need to do it numerically.  See the `getFreeEnergy()` method, which shows how the bias is stored and how the free energy is calculated:

https://github.com/openmm/openmm/blob/13b93e625c7fece95c440a2ae51cf684470ecd51/wrappers/python/openmm/app/metadynamics.py#L182-L189

Strictly speaking, the values stored in `_totalBias` are the control points of a natural (or periodic) cubic spline surface.  That's how you should interpret them if you want to calculate the integral as accurately as possible.

If you want to discuss this further, let's create a new issue for it.
